server/src/main/java/org/apache/druid/metadata/segment/cache/HeapMemorySegmentMetadataCache.java

```diff
@@ -0,0 +1,623 @@
+{
+  private static final EmittingLogger log = new EmittingLogger(HeapMemorySegmentMetadataCache.class);
+  private static final String METRIC_PREFIX = "segment/metadataCache/";
+
+  private enum CacheState
+  {
+    STOPPED, FOLLOWER, LEADER_FIRST_SYNC_PENDING, LEADER_FIRST_SYNC_STARTED, LEADER_READY
+  }
+
+  private final ObjectMapper jsonMapper;
+  private final Duration pollDuration;
+  private final boolean isCacheEnabled;
+  private final MetadataStorageTablesConfig tablesConfig;
+  private final SQLMetadataConnector connector;
+
+  private final ScheduledExecutorService pollExecutor;
+  private final ServiceEmitter emitter;
+
+  private final Object cacheStateLock = new Object();
+
+  @GuardedBy("cacheStateLock")
+  private volatile CacheState currentCacheState = CacheState.STOPPED;
+
+  private final ConcurrentHashMap<String, HeapMemoryDatasourceSegmentCache>
+      datasourceToSegmentCache = new ConcurrentHashMap<>();
+  private final AtomicReference<DateTime> syncFinishTime = new AtomicReference<>();
+
+  @Inject
+  public HeapMemorySegmentMetadataCache(
+      ObjectMapper jsonMapper,
+      Supplier<SegmentsMetadataManagerConfig> config,
+      Supplier<MetadataStorageTablesConfig> tablesConfig,
+      SQLMetadataConnector connector,
+      ScheduledExecutorFactory executorFactory,
+      ServiceEmitter emitter
+  )
+  {
+    this.jsonMapper = jsonMapper;
+    this.isCacheEnabled = config.get().isUseCache();
+    this.pollDuration = config.get().getPollDuration().toStandardDuration();
+    this.tablesConfig = tablesConfig.get();
+    this.connector = connector;
+    this.pollExecutor = isCacheEnabled ? executorFactory.create(1, "SegmentMetadataCache-%s") : null;
+    this.emitter = emitter;
+  }
+
+
+  @Override
+  @LifecycleStart
+  public void start()
+  {
+    synchronized (cacheStateLock) {
+      if (isCacheEnabled && currentCacheState == CacheState.STOPPED) {
+        currentCacheState = CacheState.FOLLOWER;
+        pollExecutor.schedule(this::syncWithMetadataStore, pollDuration.getMillis(), TimeUnit.MILLISECONDS);
+
+        log.info("Starting sync with metadata store. Cache is now in state[%s].", currentCacheState);
+      }
+    }
+  }
+
+  @Override
+  @LifecycleStop
+  public void stop()
+  {
+    synchronized (cacheStateLock) {
+      if (isCacheEnabled) {
+        pollExecutor.shutdownNow();
+        datasourceToSegmentCache.forEach((datasource, cache) -> cache.stop());
+        datasourceToSegmentCache.clear();
+
+        currentCacheState = CacheState.STOPPED;
+        log.info("Stopped sync with metadata store. Cache is now in state[%s].", currentCacheState);
+      }
+    }
+  }
+
+  @Override
+  public void becomeLeader()
+  {
+    synchronized (cacheStateLock) {
+      if (isCacheEnabled) {
+        if (currentCacheState == CacheState.STOPPED) {
+          throw DruidException.defensive("Cache has not been started yet");
+        }
+
+        currentCacheState = CacheState.LEADER_FIRST_SYNC_PENDING;
```

Comment:
Interrupt and re-start the current sync? That could help the leader gain leadership faster.

Reply:
Updated.